pkg/controllers/clusterresourceplacement/controller.go

```
@@ -324,47 +363,67 @@ func (r *Reconciler) lookupLatestClusterPolicySnapshot(ctx context.Context, crp
-	lastPolicyIndex := -1 // the assigned policy index of the cluster policy snapshot
-	for i := range snapshotList.Items {
-		policyIndex, err := parsePolicyIndexFromLabel(&snapshotList.Items[i])
+	latestSnapshot := &sortedList.Items[len(sortedList.Items)-1]
+	policyIndex, err := parsePolicyIndexFromLabel(latestSnapshot)
+	if err != nil {
+		klog.ErrorS(err, "Failed to parse the policy index label", "clusterResourcePlacement", crpKObj, "clusterPolicySnapshot", klog.KObj(latestSnapshot))
+		return nil, -1, nil, controller.NewUnexpectedBehaviorError(err)
+	}
+	return latestSnapshot, policyIndex, sortedList, nil
+}
+
+// listSortedClusterSchedulingPolicySnapshots returns the policy snapshots sorted by the policy index.
+func (r *Reconciler) listSortedClusterSchedulingPolicySnapshots(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement) (*fleetv1beta1.ClusterSchedulingPolicySnapshotList, error) {
+	snapshotList := &fleetv1beta1.ClusterSchedulingPolicySnapshotList{}
+	crpKObj := klog.KObj(crp)
+	if err := r.Client.List(ctx, snapshotList, client.MatchingLabels{fleetv1beta1.CRPTrackingLabel: crp.Name}); err != nil {
+		klog.ErrorS(err, "Failed to list all clusterSchedulingPolicySnapshots", "clusterResourcePlacement", crpKObj)
+		// CRP controller needs a scheduling policy snapshot watcher to enqueue the CRP request.
+		// So the snapshots should be read from cache.
+		return nil, controller.NewAPIServerError(true, err)
+	}
+	sort.Slice(snapshotList.Items, func(i, j int) bool {
+		ii, err := parsePolicyIndexFromLabel(&snapshotList.Items[i])
 		if err != nil {
-			klog.ErrorS(err, "Failed to parse the policy index label", "clusterPolicySnapshot", klog.KObj(&snapshotList.Items[i]))
-			return nil, -1, controller.NewUnexpectedBehaviorError(err)
+			klog.ErrorS(controller.NewUnexpectedBehaviorError(err), "Failed to parse the policy index label", "clusterSchedulingPolicySnapshot", klog.KObj(&snapshotList.Items[i]))
```

If an error occurs here, the sort result is unclear. I am not sure if simply logging the error and leave the default value for comparison is the best behavior here. It's better to have a check beforehand.

---

done, returned the unexpected error instead of error swallowing.